FWCore/Framework/src/Worker.h

```diff
@@ -723,13 +833,21 @@ namespace edm {
         });
         prePrefetchSelectionAsync(selectionTask,streamID, &ep);
       } else {
-        auto runTask = new (tbb::task::allocate_root()) RunModuleTask<T>(
-                           this, ep,es,streamID,parentContext,context);
-        prefetchAsync(runTask, parentContext, ep);
+        WaitingTask* moduleTask = nullptr;
+        if (!T::isEvent_ || !hasAcquire()) {
+          moduleTask = new (tbb::task::allocate_root()) RunModuleTask<T>(
+            this, ep, es, streamID, parentContext, context);
+        } else {
+          WaitingTaskWithArenaHolder runTaskHolder(new (tbb::task::allocate_root()) RunModuleTask<T>(
+            this, ep, es, streamID, parentContext, context));
+          moduleTask = new (tbb::task::allocate_root()) AcquireTask<T>(
+            this, ep, es, parentContext, std::move(runTaskHolder));
+        }
```

Comment:
Could this logic be simplified to
```C++
auto moduleTask = new (tbb::task::allocate_root()) RunModuleTask<T>(
                            this, ep,es,streamID,parentContext,context);
if (T::isEvent_ && hasAcquire()) {
 moduleTask = new (tbb::task::allocate_root()) AcquireTask<T>(
             this, ep, es, parentContext, edm::WaitingTaskWithArenaHolder{moduleTask});
}
prefetchAsync(moduleTask, parentContext, ep);
```

Reply:
Yes and I did it, but the same code was also affected by the changed way of handling external work exceptions.